File: src/reporter.ts
import type { Baseline, ComparatorReport, Metrics, Regression } from "./types.js";

const MARKER = "<!-- quality-gate-marker:v1 -->";

function shortSha(sha: string): string {
  return sha.slice(0, 6);
}

function fmtDelta(d: number, unit = ""): string {
  const sign = d > 0 ? "+" : "";
  return `${sign}${d}${unit}`;
}

function row(name: string, baseline: string, pr: string, delta: string, status: string): string {
  return `| ${name.padEnd(13)} | ${baseline.padEnd(9)} | ${pr.padEnd(7)} | ${delta.padEnd(7)} | ${status.padEnd(7)} |`;
}

function buildTable(pr: Metrics, baseline: Baseline | null, report: ComparatorReport): string {
  const bl = baseline?.metrics;
  const failed = new Set(report.regressions.map(r => r.metric));
  const rows: string[] = [
    row("Metric", "Baseline", "PR", "Δ", "Status"),
    "|---------------|----------|---------|---------|---------|",
  ];

  if ("lines_pct" in pr.coverage) {
    const b = bl && "lines_pct" in bl.coverage ? bl.coverage.lines_pct : null;
    const p = pr.coverage.lines_pct;
    const d = b === null ? "—" : `${(p - b).toFixed(2)}pp`;
    rows.push(row("Coverage", b !== null ? `${b}%` : "—", `${p}%`, d, failed.has("coverage") ? "❌" : "✅"));
  }
  if ("pct" in pr.duplication) {
    const b = bl && "pct" in bl.duplication ? bl.duplication.pct : null;
    const p = pr.duplication.pct;
    const d = b === null ? "—" : `${(p - b).toFixed(2)}pp`;
    rows.push(row("Duplication", b !== null ? `${b}%` : "—", `${p}%`, d, failed.has("duplication") ? "❌" : "✅"));
  }
  if ("total" in pr.lint) {
    const b = bl && "total" in bl.lint ? bl.lint.total : null;
    const p = pr.lint.total;
    const d = b === null ? "—" : fmtDelta(p - b);
    rows.push(row("Lint (total)", b !== null ? `${b}` : "—", `${p}`, d, failed.has("lint") ? "❌" : "✅"));
  }
  if ("violations" in pr.file_size) {
    const b = bl && "violations" in bl.file_size ? Object.keys(bl.file_size.violations).length : null;
    const p = Object.keys(pr.file_size.violations).length;
    const d = b === null ? "—" : fmtDelta(p - b);
    rows.push(row("Oversized", b !== null ? `${b}` : "—", `${p}`, d, failed.has("file_size") ? "❌" : "✅"));
  }
  if ("critical" in pr.security) {
    const b = bl && "critical" in bl.security ? bl.security.critical : null;
    rows.push(row("Security", b !== null ? `${b} crit` : "—", `${pr.security.critical} crit`, "—",
      report.regressions.some(r => r.metric === "security") ? "❌" : "✅"));
  }
  return rows.join("\n");
}

function regressionSections(report: ComparatorReport): string {
  if (report.regressions.length === 0) return "";
  const parts = ["### Regressions\n"];
  const byMetric = new Map<string, Regression[]>();
  for (const r of report.regressions) {
    if (!byMetric.has(r.metric)) byMetric.set(r.metric, []);
    byMetric.get(r.metric)!.push(r);
  }
  for (const metric of Array.from(byMetric.keys()).sort()) {
    const items = byMetric.get(metric)!;
    const lines = [`**${metric}** — ${items.length} issue(s):`];
    const sorted = [...items].sort((a, b) =>
      (a.scope ?? "").localeCompare(b.scope ?? "") || (a.file ?? "").localeCompare(b.file ?? ""),
    );
    for (const it of sorted) {
      const filePart = it.file ? ` \`${it.file}\`` : "";
      let extra = "";
      if (it.actual !== undefined && it.baseline !== undefined) {
        extra = ` ${it.baseline} → ${it.actual} (${fmtDelta(it.delta ?? 0)})`;
      } else if (it.actual !== undefined && it.required !== undefined) {
        extra = ` ${it.actual} (required: ${it.required})`;
      }
      lines.push(`- [${it.scope ?? "global"}]${filePart}${extra}`);
    }
    parts.push(lines.join("\n"), "");
  }
  return parts.join("\n");
}

function warningSection(report: ComparatorReport): string {
  if (report.warnings.length === 0) return "";
  const lines = ["### Warnings (non-blocking)\n"];
  for (const w of report.warnings) {
    lines.push(`- **${w.metric}**: ${w.count} \`${w.severity}\` (Δ ${fmtDelta(w.delta ?? 0)})`);
  }
  return lines.join("\n") + "\n";
}

export function renderPrComment(pr: Metrics, baseline: Baseline | null, report: ComparatorReport): string {
  const adapter = `${pr._meta.adapter}@${pr._meta.adapter_version}`;
  const tools = pr._meta.tools.join(", ");

  let header: string;
  let baselineLine: string;
  if (report.bootstrap) {
    header = "⚙️ BOOTSTRAP";
    baselineLine = "> Baseline: _pending — will be created on first merge to default branch._";
  } else {
    header = report.gate_passed ? "✅ PASSED" : "❌ FAILED";
    baselineLine = baseline ? `> Baseline: \`${shortSha(baseline.commit_sha)}\` (${baseline.updated_at.slice(0, 10)})` : "";
  }

  const parts: string[] = [
    MARKER,
    `## 🚦 Quality Gate — ${header}`,
    "",
    `> Adapter: \`${adapter}\` · Tools: ${tools}`,
  ];
  if (baselineLine) parts.push(baselineLine);
  parts.push("", buildTable(pr, baseline, report), "");
  if (report.bootstrap) {
    parts.push("⚙️ **Bootstrap pending** — baseline will be created on first merge to default branch. Critical security issues still block.\n");
  }
  parts.push(regressionSections(report));
  parts.push(warningSection(report));
  return parts.filter(p => p !== null).join("\n").replace(/\n+$/, "") + "\n";
}

const COVERAGE_THRESHOLDS: Array<[number, string]> = [[80, "brightgreen"], [60, "yellowgreen"], [40, "yellow"], [20, "orange"], [0, "red"]];
const DUPLICATION_THRESHOLDS: Array<[number, string]> = [[2, "brightgreen"], [5, "yellow"], [10, "orange"], [1000, "red"]];
const LINT_THRESHOLDS: Array<[number, string]> = [[0, "brightgreen"], [10, "yellow"], [50, "orange"], [1_000_000, "red"]];

function coverageColor(pct: number): string {
  for (const [b, c] of COVERAGE_THRESHOLDS) if (pct >= b) return c;
  return "red";
}

function duplicationColor(pct: number): string {
  for (const [b, c] of DUPLICATION_THRESHOLDS) if (pct <= b) return c;
  return "red";
}

function lintColor(total: number): string {
  for (const [b, c] of LINT_THRESHOLDS) if (total <= b) return c;
  return "red";
}

function shields(label: string, message: string, color: string): string {
  return JSON.stringify({ schemaVersion: 1, label, message, color }, null, 2) + "\n";
}

export function renderBadges(pr: Metrics): Record<string, string> {
  if (!("lines_pct" in pr.coverage) || !("pct" in pr.duplication) || !("total" in pr.lint)) {
    throw new Error("renderBadges requires non-skipped coverage, duplication, and lint metrics");
  }
  const covColor = coverageColor(pr.coverage.lines_pct);
  const dupColor = duplicationColor(pr.duplication.pct);
  const lintC = lintColor(pr.lint.total);
  const rank: Record<string, number> = { brightgreen: 0, yellowgreen: 1, yellow: 2, orange: 3, red: 4 };
  const composite = [covColor, dupColor, lintC].reduce((a, b) => ((rank[a] ?? 4) >= (rank[b] ?? 4) ? a : b));
  return {
    "coverage.json":    shields("coverage", `${pr.coverage.lines_pct}%`, covColor),
    "duplication.json": shields("duplication", `${pr.duplication.pct}%`, dupColor),
    "lint.json":        shields("lint", String(pr.lint.total), lintC),
    "quality.json":     shields("quality", composite === "brightgreen" ? "passing" : "issues", composite),
  };
}

File: qg_core/reporter.py
"""Reporter: comparator-report.json + metrics + baseline -> markdown for PR + summary."""
from __future__ import annotations

from typing import Any

MARKER = "<!-- quality-gate-marker:v1 -->"


def _short_sha(sha: str) -> str:
    return sha[:6]


def _fmt_delta(delta: float | int, unit: str = "") -> str:
    sign = "+" if delta > 0 else ""
    return f"{sign}{delta}{unit}"


def _table_row(name: str, baseline_val: str, pr_val: str, delta: str, status: str) -> str:
    return f"| {name:<14}| {baseline_val:<9}| {pr_val:<8}| {delta:<7}| {status:<7}|"


def _build_table(pr: dict, baseline: dict | None, report: dict) -> str:
    bl = baseline["metrics"] if baseline else {}
    failed_metrics = {r["metric"] for r in report["regressions"]}

    rows: list[str] = []
    rows.append(_table_row("Metric", "Baseline", "PR", "Δ", "Status"))
    rows.append("|---------------|----------|--------|--------|--------|")

    if "coverage" in pr and "lines_pct" in pr["coverage"]:
        b = bl.get("coverage", {}).get("lines_pct") if bl else None
        p = pr["coverage"]["lines_pct"]
        d = "—" if b is None else f"{p - b:+.2f}pp"
        rows.append(_table_row("Coverage", f"{b}%" if b is not None else "—", f"{p}%", d,
                               "❌" if "coverage" in failed_metrics else "✅"))

    if "duplication" in pr and "pct" in pr["duplication"]:
        b = bl.get("duplication", {}).get("pct") if bl else None
        p = pr["duplication"]["pct"]
        d = "—" if b is None else f"{p - b:+.2f}pp"
        rows.append(_table_row("Duplication", f"{b}%" if b is not None else "—", f"{p}%", d,
                               "❌" if "duplication" in failed_metrics else "✅"))

    if "lint" in pr and "total" in pr["lint"]:
        b = bl.get("lint", {}).get("total") if bl else None
        p = pr["lint"]["total"]
        d = "—" if b is None else _fmt_delta(p - b)
        rows.append(_table_row("Lint (total)", str(b) if b is not None else "—", str(p), d,
                               "❌" if "lint" in failed_metrics else "✅"))

    if "file_size" in pr and "violations" in pr["file_size"]:
        b = len(bl.get("file_size", {}).get("violations", {})) if bl else None
        p = len(pr["file_size"]["violations"])
        d = "—" if b is None else _fmt_delta(p - b)
        rows.append(_table_row("Oversized", str(b) if b is not None else "—", str(p), d,
                               "❌" if "file_size" in failed_metrics else "✅"))

    if "security" in pr and "critical" in pr["security"]:
        p = pr["security"]["critical"]
        b = bl.get("security", {}).get("critical") if bl else None
        rows.append(_table_row("Security", f"{b} crit" if b is not None else "—",
                               f"{p} crit", "—",
                               "❌" if any(r["metric"] == "security" for r in report["regressions"]) else "✅"))
    return "\n".join(rows)


def _regression_sections(report: dict) -> str:
    if not report["regressions"]:
        return ""
    parts = ["### Regressions\n"]
    by_metric: dict[str, list[dict]] = {}
    for r in report["regressions"]:
        by_metric.setdefault(r["metric"], []).append(r)
    for metric in sorted(by_metric):
        items = by_metric[metric]
        lines = [f"**{metric}** — {len(items)} issue(s):"]
        for it in sorted(items, key=lambda x: (x.get("scope", ""), x.get("file", ""))):
            file_part = f" `{it['file']}`" if "file" in it else ""
            extra = ""
            if "actual" in it and "baseline" in it:
                extra = f" {it['baseline']} → {it['actual']} ({_fmt_delta(it.get('delta', 0))})"
            elif "actual" in it and "required" in it:
                extra = f" {it['actual']} (required: {it['required']})"
            lines.append(f"- [{it.get('scope', 'global')}]{file_part}{extra}")
        parts.append("\n".join(lines))
        parts.append("")
    return "\n".join(parts)


def _warning_section(report: dict) -> str:
    if not report["warnings"]:
        return ""
    lines = ["### Warnings (non-blocking)\n"]
    for w in report["warnings"]:
        lines.append(f"- **{w['metric']}**: {w['count']} `{w['severity']}` (Δ {_fmt_delta(w.get('delta', 0))})")
    return "\n".join(lines) + "\n"


def render_pr_comment(pr_metrics: dict, baseline: dict | None, report: dict) -> str:
    meta = pr_metrics.get("_meta", {})
    adapter = f"{meta.get('adapter', '?')}@{meta.get('adapter_version', '?')}"
    tools = ", ".join(meta.get("tools", []))

    if report.get("bootstrap"):
        header = "⚙️ BOOTSTRAP"
        baseline_line = "> Baseline: _pending — will be created on first merge to default branch._"
    else:
        header = "✅ PASSED" if report["gate_passed"] else "❌ FAILED"
        if baseline is not None:
            baseline_line = f"> Baseline: `{_short_sha(baseline['commit_sha'])}` ({baseline['updated_at'][:10]})"
        else:
            baseline_line = ""

    parts = [
        MARKER,
        f"## 🚦 Quality Gate — {header}",
        "",
        f"> Adapter: `{adapter}` · Tools: {tools}",
        baseline_line,
        "",
        _build_table(pr_metrics, baseline, report),
        "",
    ]
    if report.get("bootstrap"):
        parts.append("⚙️ **Bootstrap pending** — baseline will be created on first merge to default branch. "
                     "Critical security issues still block.\n")
    parts.append(_regression_sections(report))
    parts.append(_warning_section(report))
    return "\n".join(p for p in parts if p is not None).rstrip() + "\n"
